File: androidTest/java/org/aprsdroid/app/DatagramRecorderSocket.java
package org.aprsdroid.app;

import android.util.Log;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketAddress;
import java.net.SocketException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

class DatagramRecorderSocket extends DatagramSocket {
    private InetAddress remote_addr = null;
    private int remote_port = 0;

    public DatagramRecorderSocket() throws SocketException {
    }

    @Override
    public void bind(SocketAddress addr) {
    }

    @Override
    public void close() {
    }

    @Override
    public void connect(SocketAddress addr) {
//        remote_addr = addr;
    }

    @Override
    public void connect(InetAddress addr, int port) {
        remote_addr = addr;
        remote_port = port;
    }

    @Override
    public void disconnect() {
    }

    @Override
    public void send(DatagramPacket p) {
        if(p != null) {
            Log.d("APRSdroid-test", "Packet to " + p.getAddress().toString() + ":" + p.getPort() + " -> " + new String(p.getData(), StandardCharsets.UTF_8));
            DatagramLog.addPacket(p);
        }
    }

    static class DatagramLog {
        private static ArrayList<DatagramPacket> log = new ArrayList<>();

        static void clear() {
            log.clear();
        }

        static void addPacket(DatagramPacket packet) {
            log.add(packet);
        }

        static List<DatagramPacket> getLog() {
            return log;
        }
    }
}

File: androidTest/java/org/aprsdroid/app/DatagramRecorderSocket.kt
package org.aprsdroid.app

import android.util.Log
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.SocketAddress

class DatagramRecorderSocket: DatagramSocket() {
    private var remote_addr: InetAddress? = null
    private var remote_port: Int = 0

    override fun bind(addr: SocketAddress?) {
    }

    override fun close() {
    }

    override fun connect(addr: SocketAddress?) {
//        remote_addr = addr
    }

    override fun connect(addr: InetAddress?, port: Int) {
        remote_addr = addr
        remote_port = port
    }

    override fun disconnect() {
    }

    override fun send(p: DatagramPacket?) {
        p?.let {
            Log.d("APRSTest", "Packet to ${it.address.toString()}:${it.port} -> ${it.data.toString()}")
        }
        if(p != null)
            DatagramLog.log.add(p)
    }
}

object DatagramLog {
    val log = arrayListOf<DatagramPacket>()

    fun clear() {
        log.clear()
    }

    fun getLog(): List<DatagramPacket> {
        return log
    }
}
